dotnet/samples/Demos/VectorStoreRAG/RAGChatService.cs

```diff
@@ -0,0 +1,166 @@
+    IDataLoader dataLoader,
+    VectorStoreTextSearch<TextSnippet<TKey>> vectorStoreTextSearch,
+    Kernel kernel,
+    IOptions<RagConfig> ragConfigOptions) : IHostedService
+{
+    private Task? _dataLoaded;
+    private Task? _chatLoop;
+
+    /// <summary>
+    /// Start the service.
+    /// </summary>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests.</param>
+    /// <returns>An async task that completes when the service is started.</returns>
+    public Task StartAsync(CancellationToken cancellationToken)
+    {
+        // Start to load all the configured PDFs into the vector store.
+        if (ragConfigOptions.Value.BuildCollection)
+        {
+            this._dataLoaded = this.LoadDataAsync(cancellationToken);
+        }
+        else
+        {
+            this._dataLoaded = Task.CompletedTask;
+        }
+
+        // Start the chat loop.
+        this._chatLoop = this.ChatLoopAsync(cancellationToken);
+
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Stop the service.
+    /// </summary>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests.</param>
+    /// <returns>An async task that completes when the service is stopped.</returns>
+    public Task StopAsync(CancellationToken cancellationToken)
+    {
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Contains the main chat loop for the application.
+    /// </summary>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests.</param>
+    /// <returns>An async task that completes when the chat loop is shut down.</returns>
+    private async Task ChatLoopAsync(CancellationToken cancellationToken)
+    {
+        // Wait for the data to be loaded before starting the chat loop.
+        while (this._dataLoaded != null && !this._dataLoaded.IsCompleted && !cancellationToken.IsCancellationRequested)
+        {
+            await Task.Delay(1_000, cancellationToken).ConfigureAwait(false);
+        }
+
+        // If data loading failed, don't start the chat loop.
+        if (this._dataLoaded != null && this._dataLoaded.IsFaulted)
+        {
+            Console.WriteLine("Failed to load data");
+            return;
+        }
+
+        Console.WriteLine("PDF loading complete\n");
+
+        // Add a search plugin to the kernel which we will use in the template below
+        // to do a vector search for related information to the user query.
+        kernel.Plugins.Add(vectorStoreTextSearch.CreateWithGetTextSearchResults("SearchPlugin"));
+
+        // Start the chat loop.
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            // Prompt the user for a question.
+            Console.ForegroundColor = ConsoleColor.Green;
+            Console.WriteLine("Assistant > What would you like to know from the PDFs?");
```

Comment:
Can we include the names of the PDF's, might help set the context

Reply:
Done.